### ISSUE TITLE
Preventing being ran over by other mods

### ISSUE THREAD

**Issue by Lisias** [net-lisias-ksp/TweakScale#31]
This isn't exact an issue but an Task.

Since some more Mods are currently mangling GameDatabase on the Main Menu Scene, and since TweakScale must be last of the Manglers on that event chain, and since TweakScale **needs** to have their internal affairs in order **before** the user loads a savegame (what's inevitably happens before loading the Space Center!) due the prefab being applied too on craft from savegames (novelty on 1.5.1 **or** 1.6.1, last time I checked it I was playing 1.4.x).

**Comment by Lisias** [comment]
Latest commit on this effort: https://github.com/net-lisias-ksp/TweakScale/commit/b885e54ba39af93f912488c8d54b2fc52edaa1d9

**Comment by Lisias** [comment]
A Beta Release to check if I didn't made any serious mistake - as well to see if by luck this really fixes something.

## WARNING

**This can break your KSP, ruin your Windows, kill your pet and poison your kids. :D** 

By the Holy Kerbol that enlighten us, please use this only under my instructions, and only if I ask you to do so! Twice. :) 

[TweakScale-Beta-2.4.1.1.zip](https://github.com/net-lisias-ksp/TweakScale/files/2975395/TweakScale-Beta-2.4.1.1.zip)

**Comment by Lisias** [comment]
This stunt apparently managed to stick - at least, it didn't make things worse.

Additionally, @Jammer-TD test runs on https://github.com/net-lisias-ksp/TweakScale/issues/42#issuecomment-490719794 suggests that the Prefab post initialization instrumenting is now able to survive Mission History, ScanSAT and KIS:

```
[LOG 20:21:08.497] TweakScale::WriteDryCost: Started
[LOG 20:21:08.498] [ModuleManager] Total loading Time = 243.598s
[LOG 20:21:08.920] ADDON BINDER: Create binding redirect: ClickThroughBlocker, Version=0.1.6.10, Culture=neutral, PublicKeyToken=null => ClickThroughBlocker, Version=0.1.7.2, Culture=n
[LOG 20:21:08.922] Tweak 'OnOrbitRightKerbal' matched kerbal model: [Transform:OrbitScene/Kerbals/maleEVA_inverted]
[LOG 20:21:08.923] Equipped part on kerbal model in main screen: KIS.bomb1
[LOG 20:21:08.923] Equipped part on kerbal model in main screen: KIS.Container1
[LOG 20:21:08.923] Tweak 'OnOrbitLeftKerbal' matched kerbal model: [Transform:OrbitScene/Kerbals/maleEVA_side]
[LOG 20:21:08.924] Equipped part on kerbal model in main screen: ScienceBox
[LOG 20:21:08.924] Tweak 'OnOrbitMiddleKerbal' matched kerbal model: [Transform:OrbitScene/Kerbals/maleEVA_center]
[LOG 20:21:08.924] Equipped part on kerbal model in main screen: KIS.electricScrewdriver
[LOG 20:21:08.924] Tweak 'OnOrbitFemaleKerbal' matched kerbal model: [Transform:OrbitScene/Kerbals/femaleEVA]
[LOG 20:21:08.924] Equipped part on kerbal model in main screen: KIS.wrench
[LOG 20:21:08.925] Equipped part on kerbal model in main screen: KIS.guide
[LOG 20:21:08.925] Tweak 'OnMunLoneKerbal' matched kerbal model: [Transform:MunScene/Kerbals/maleEVA]
[LOG 20:21:08.925] Equipped part on kerbal model in main screen: KIS.electricScrewdriver
[LOG 20:21:08.925] Equipped part on kerbal model in main screen: KIS.Container1
[LOG 20:21:08.932] [ScenarioTypes]: List Created 24 scenario types loaded from 55 loaded assemblies.
[LOG 20:21:08.938] [ApplicationLauncher] Awake False
[LOG 20:21:08.939] [ApplicationLauncher] SpawnSimpleLayout: VerticalTopDown
[LOG 20:21:08.949] ScaleModList: listSize 41 maxListSize 0
[LOG 20:21:08.952] [UiApp] Awake: MessageSystem
[LOG 20:21:08.953] [ApplicationLauncher] OnSceneLoadedGUIReady: scene MAINMENU ShouldBeVisible() True ShouldBeOnTop() True iIsPositionedAtTop True
[LOG 20:21:08.961] [SCANsat] Shader asset bundle loaded; using platform bundle: scan_shaders-windows.scan
[LOG 20:21:08.966] [SCANsat] Unity skin asset bundle loaded
[LOG 20:21:08.969] [SCANsat] Icon asset bundle loaded
[LOG 20:21:09.073] [SCANsat] UI prefab bundle loaded and processed
[LOG 20:21:09.073] [SCANsat] All SCANsat asset bundles loaded
[LOG 20:21:09.122] IRS: Successfully loaded all prefabs from AssetBundle
[WRN 20:21:09.162] [TweakScale] Removing TweakScale support for FSfloatEnd.
[ERR 20:21:09.162] [TweakScale] Part FSfloatEnd didn't passed the sanity check due using FSbuoyancy module - see issue #9 https://github.com/net-lisias-ksp/TweakScale/issues/9.

[WRN 20:21:09.162] [TweakScale] Removing TweakScale support for FSfloatEndTail.
[ERR 20:21:09.162] [TweakScale] Part FSfloatEndTail didn't passed the sanity check due using FSbuoyancy module - see issue #9 https://github.com/net-lisias-ksp/TweakScale/issues/9.

[WRN 20:21:09.162] [TweakScale] Removing TweakScale support for FSfloatGearbay.
[ERR 20:21:09.162] [TweakScale] Part FSfloatGearbay didn't passed the sanity check due using FSbuoyancy module - see issue #9 https://github.com/net-lisias-ksp/TweakScale/issues/9.

[WRN 20:21:09.162] [TweakScale] Removing TweakScale support for FSfloatStraight.
[ERR 20:21:09.162] [TweakScale] Part FSfloatStraight didn't passed the sanity check due using FSbuoyancy module - see issue #9 https://github.com/net-lisias-ksp/TweakScale/issues/9.

[WRN 20:21:09.162] [TweakScale] Removing TweakScale support for FSfloatStrut.
[ERR 20:21:09.162] [TweakScale] Part FSfloatStrut didn't passed the sanity check due using FSbuoyancy module - see issue #9 https://github.com/net-lisias-ksp/TweakScale/issues/9.

[WRN 20:21:09.162] [TweakScale] Removing TweakScale support for FSlandingPads.
[ERR 20:21:09.162] [TweakScale] Part FSlandingPads didn't passed the sanity check due using FSbuoyancy module - see issue #9 https://github.com/net-lisias-ksp/TweakScale/issues/9.

[ERR 20:21:09.162] TweakScale::PrefabDryCostWriter: negative dryCost: part=IFSHexcanSolid, DryCost=-1.66893E-05

[LOG 20:21:09.166] TweakScale::WriteDryCost: Concluded
```

(i.e. - the stunt apparently is detecting when something else is mangling the GameDatabase and manages to delay the task!)

So I'm closing this as FIXED. for while :)

**Comment by Lisias** [comment]
It's happening again. =/

**Comment by Lisias** [comment]
I got news that Pre Releaser 2.5.0.2 had this problem solved on the field for at least one user.

See this [Post on TweakScale's Thread](https://forum.kerbalspaceprogram.com/index.php?/topic/179030-14-tweakscale-under-lisias-management-2430-2019-0608/&do=findComment&comment=3629378) .

**Comment by Lisias** [comment]
This time the problem was another one. :)

Fixed on commit: 

[Source](https://forum.kerbalspaceprogram.com/index.php?/topic/7542-the-official-unoffical-quothelp-a-fellow-plugin-developerquot-thread/&do=findComment&comment=3631853)

**Comment by Lisias** [comment]
Well… Brother In Arms LinuxGuruGamer [reported on Forum](https://forum.kerbalspaceprogram.com/index.php?/topic/179030-14-tweakscale-under-lisias-management-2433-2019-0814/&do=findComment&comment=3664500) a new situation where this is happening.

Well.. Let the fun begins. **AGAIN**. :P 

[TSErrorLogs-2ndSet.zip](https://github.com/net-lisias-ksp/TweakScale/files/3563017/TSErrorLogs-2ndSet.zip)

**Comment by linuxgurugamer** [comment]
Not sure I understand the issue, can you  explain it please

**Comment by Lisias** [comment]
I'm finishing a PhD thesis to be published on the Forum about some issues, and this is one of them - believe it or not!

Check the forum in the next few minutes. I'm cooking a binary to be executed on your installment to see what happens. If this stunt stick, we need to figure out why it worked before I can write a proper patch to this workaound.

**Comment by Lisias** [comment]
For the record, that PhD thesis is [here](https://forum.kerbalspaceprogram.com/index.php?/topic/179030-14-tweakscale-under-lisias-management-2433-2019-0814/&do=findComment&comment=3664640).

**Comment by Lisias** [comment]
@linuxgurugamer , please download this ZIP and unpack it on the Plugins folder on the GameData/TweakScale - be sure to overwrite (or delete) any DLL there.

Then fire up the thing.

This DLL will impose itself a 15 seconds moratory before trying to do its business. A pesky AlertBox will appears during the whole process (moratory, DryCost patching, Sanity Checks).

Do not load any savegame, just quit KSP and send me the KSP.log to see if any NRE were issued. I hope not.

With this confirmation, the next step is to cook some heuristics in order to fine tune that moratory. The mainstream DLL has a moratory of 0 to 2 seconds max, needed by Making History - Serenity apparently doesn't impose an additional tax on the thing.

I don't know how ScanSAT and KISS handle this (time to make some exploratory digging)

Binary:  [Plugins.zip](https://github.com/net-lisias-ksp/TweakScale/files/3563079/Plugins.zip)
Source: https://github.com/net-lisias-ksp/TweakScale/tree/issue/31

**Comment by linuxgurugamer** [comment]
ok, ill try it

**Comment by linuxgurugamer** [comment]
eegads!  What are you doing to my poor laptop?  This is an i7-7700HQ, it's running at nearly 75%cpu.

The dialog says to wait, but I'm seeing the errors in the log file already.  I'llgive it another few minutes and then kill it:


[TweakScale] ERROR: part=bluedog.Titan2.1875mDecoupler (Prometheus-II 1.875m Stack Decoupler) Exception on Sanity Checks: System.NullReferenceException: Object reference not set to an instance of an object
  at PartModuleList.Contains (Int32 classID) [0x00000] in <filename unknown>:0 
  at PartModuleList.Contains (System.String className) [0x00000] in <filename unknown>:0 
  at TweakScale.PrefabDryCostWriter.checkForSanity (.Part p) [0x00000] in <filename unknown>:0 
  at TweakScale.PrefabDryCostWriter+<WriteDryCost>d__3.MoveNext () [0x00000] in <filename unknown>:0 
 
(Filename: C:/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)

[TweakScale] ERROR: part=bluedog.Titan2.1875mDecoupler (Prometheus-II 1.875m Stack Decoupler) Exception on writeDryCost: System.NullReferenceException: Object reference not set to an instance of an object
  at PartModuleList.Contains (Int32 classID) [0x00000] in <filename unknown>:0 
  at PartModuleList.Contains (System.String className) [0x00000] in <filename unknown>:0 
  at TweakScale.PrefabDryCostWriter+<WriteDryCost>d__3.MoveNext () [0x00000] in <filename unknown>:0 
 
(Filename: C:/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)

[TweakScale] ERROR: part=bluedog.Titan2.FuelTank1 (Prometheus-II-3000 Liquid Fuel Tank) Exception on Sanity Checks: System.NullReferenceException: Object reference not set to an instance of an object
  at PartModuleList.Contains (Int32 classID) [0x00000] in <filename unknown>:0 
  at PartModuleList.Contains (System.String className) [0x00000] in <filename unknown>:0 
  at TweakScale.PrefabDryCostWriter.checkForSanity (.Part p) [0x00000] in <filename unknown>:0 
  at TweakScale.PrefabDryCostWriter+<WriteDryCost>d__3.MoveNext () [0x00000] in <filename unknown>:0

**Comment by linuxgurugamer** [comment]
I stopped it before it melted my laptop.
Here are all the files:
https://www.dropbox.com/s/a8j5e1c41a85ho6/TS-3.zip?dl=0

**Comment by linuxgurugamer** [comment]
It's 9:30 pm here, I'll be on for another 1/2 hour, and then to sleep, tomorrow am flying home

**Comment by Lisias** [comment]
> eegads! What are you doing to my poor laptop? This is an i7-7700HQ, it's running at nearly 75%cpu.
> 
> The dialog says to wait, but I'm seeing the errors in the log file already. I'llgive it another few minutes and then kill it:
 
Interesting. That was just a loop on a co-routine that just yield 900 times, one time a frame.

No waits, not sleeps. Just a yield.

And, of course, I tested the thing on my MacPotato.

The `[LOG 21:11:20.348] [TweakScale] WARNING: Timeout waiting for PartLoader.LoadedPartsList!!` log entry hints that this wait has reached the end alright.

There're no subsequent TweakScake Timeout warning on the log, so this implies that the second loop decided not to run. And yet, we have this enormous amount of NREs. Essentially **for every single part** on the installment, as it appears!

So whatever is mangling `PartLoader.LoadedPartsList` and breaking TweakScale, had done it way before. But this means that the `partPrefab` of each part in the GameDatabase is corrupted by the time TweakScale tries to inspect it.

Damn, dude. This is way worse than simply a race condition!

**Comment by Lisias** [comment]
Found the reason for the notebook's Syndrome of China. I didn't had foresee this exact situation, i.e., a completely new `PartLoader.LoadedPartsList` out from the blue. So the code stubbornly retried and retried the same stunt on the same (invalid) data. Of course I had yielded the co-routine control on each failure, but once TweakScale is the only code running, the yield ends up doing nothing significant.

Cooking a new binary.

**Comment by Lisias** [comment]
New try. Don't let it run for more than 5 minutes if it gets stuck.

Binary: [Plugins.zip](https://github.com/net-lisias-ksp/TweakScale/files/3563730/Plugins.zip)
Source: https://github.com/net-lisias-ksp/TweakScale/tree/issue/31

**Comment by linuxgurugamer** [comment]
Looks much better, will provide a set of files soon

**Comment by linuxgurugamer** [comment]
Here you go.  Still nullrefs, but it did stop after about 4 minutes
https://www.dropbox.com/s/r4lpgjnn7mctf0j/TS-Logs-4.zip?dl=0

**Comment by Lisias** [comment]
Without that long moratory stunt, the process did managed to get to the end.

```
[LOG 16:44:22.272] [TweakScale] INFO: TweakScale::WriteDryCost: Concluded : 2172 checks failed ; 0 parts with hotfixes ; 0 parts with issues overruled ; 0 Show Stoppers found; 9 Sanity Check failed;
```

Other than the NSS/Parts/Update parts present in the ConfigCache of this new log and you had used your desktop this time, everything else is precisely identical - both test runs appears to be similar (except by the need of aborting the previous one due the borked moratory).

From this log, I realized that not all parts are borking! The parts that borked on the last log still borks on this, so that new stunt of mine (reworking how the PartLoader's parts are fetched) didn't accomplish nothing. However, this new way assures to get always a current data from the Database, as I'm not using a foreach anymore (that prebuilds a list of parts to be iterated early on the start).

So I think that no, the PartLoader''s database is not being mangled at the same time TweakScale is trying to inspect this time - whatever had happened to the database, must be already done by the time TweakScale tries to inspect it.

Now I need to figure out who is the one doing that, and why the PartLoader database is a wreck due that stunt.

**Comment by Lisias** [comment]
@linuxgurugamer got out of time this week, my next time window will be Sunday. I'm still on it!

**Comment by linuxgurugamer** [comment]
not a problem, this is in a new install I'm setting up, am just not installing TS for now

**Comment by Lisias** [comment]
Well… At least I know one more thing that **DOES NOT** affects TweakScale. ReStock and RestockPlus appears to doing everything by the book.

I decided to check this first as this is something that heavily mangles the GameDatabase, so the surface of exposition is huge. But apparently this is doing fine.

This is going to be worksome. :P

**Comment by Lisias** [comment]
Waving the milestone. This is a pressuring issue, but it was run over by yet more pressuring issues - and now I had run out of enough time to give this a new round.

**Comment by AccidentalDisassembly** [comment]
Hi Lisias - log as requested from the forum, re: something going wonky at main menu (maybe?). 
Log is here: https://www.dropbox.com/s/alslbgc39nirm8f/KSPLog_TweakScale_5_2_2020.log?dl=0 
Forum post is here: https://forum.kerbalspaceprogram.com/index.php?/topic/179030-ksp-141-tweakscale-under-lisias-management-24312-2020-0426/&do=findComment&comment=3779758

**Comment by Lisias** [comment]
This task was closed by side effect while implementing #256

**Comment by Lisias** [comment]
Screw that. This is going gold into 2.5.0.0
